programs/mango-v4/src/instructions/token_charge_collateral_fees.rs

```diff
@@ -0,0 +1,77 @@
+use crate::accounts_zerocopy::*;
+use crate::error::*;
+use crate::health::*;
+use crate::state::*;
+use anchor_lang::prelude::*;
+use anchor_spl::associated_token;
+use anchor_spl::token;
+use fixed::types::I80F48;
+
+use crate::accounts_ix::*;
+use crate::logs::{
+    emit_stack, LoanOriginationFeeInstruction, TokenBalanceLog, WithdrawLoanLog, WithdrawLog,
+};
+
+pub fn token_charge_collateral_fees(ctx: Context<TokenChargeCollateralFees>) -> Result<()> {
+    let group = ctx.accounts.group.load()?;
+    if group.collateral_fee_interval == 0 {
+        return Ok(());
+    }
+
+    let mut account = ctx.accounts.account.load_full_mut()?;
+    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+
+    // TODO: should we charge collateral fees for accounts that are being liquidated?
+    require!(!account.being_liquidated(), MangoError::BeingLiquidated);
+
+    // Is the next fee-charging due?
+    let last_charge_ts = account.fixed.last_collateral_fee_charge;
+    if now_ts < last_charge_ts + group.collateral_fee_interval {
+        return Ok(());
+    }
+    account.fixed.last_collateral_fee_charge = now_ts;
+
+    // Charge the user at most for 2x the interval. So if no one calls this for a long time
+    // there won't be a huge charge based only on the end state.
+    let charge_seconds = (now_ts - last_charge_ts).min(2 * group.collateral_fee_interval);
+
+    let inv_seconds_per_day = I80F48::from_num(1.157407407407e-5); // 1 / (24 * 60 * 60)
+    let collateral_fee_scaling = I80F48::from(charge_seconds) * inv_seconds_per_day;
```

Comment:
q: why this multiplication with inv? what is the usefulness of the inv?

Reply:
I assumed we configure the fees per day (similar to funding). So if we run this instruction once every 12-24h (example) then we need to compute how much of the daily fees apply each time. So it's `secs_since_last_time / seconds_per_day` to get a fraction of a day